rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,572 @@
+6. If a meaningful defaults exist for fields on dev commands, that field should be modelled as an `Option`.
+   1. This allows CLI-style toolboxes to more easily populate your struct.
+
+### Building toolboxes
+
+Let's take a look at how this all fits together by attempting to build a Quake-style dev console, much like [`bevy-console`](https://github.com/RichoDemus/bevy-console).
+Obviously, this is too big to cover properly in an RFC example, so we're going to hand-wave the user input side of things by showing how you might deal with the parsed events, and simplify output by just logging to `println!`.
+
+Our first task is getting the list of dev tools.
+
+```rust
+#[derive(Event)]
+struct ListDevTools;
+
+fn list_dev_tools(mut event_reader: EventReader<ListDevTools>, dev_tools_registry: Res<DevToolsRegistry>){
+    // Clear the events, and act if at least one is found
+    if event_reader.drain().next().is_some() {
+        println!("Available modal dev tools:");
+        for (_component_id, modal_dev_tool) in dev_tools_registry.iter_modal_dev_tools() {
+            println!("{}", modal_dev_tool.type_name());
+            println!("{}", modal_dev_tool.docs());
+
+            for field in modal_dev_tools.fields(){
+                println!("{}", field.type_name());
+                println!("{}", field.docs())
+            }
+        }
+    }
+}
+```
+
+Next, we want to be able to toggle each modal tool by name.
+
+```rust
+#[derive(Event)]
+struct ToggleDevTool{
+    name: String,
+}
+
+fn toggle_dev_tools(world: &mut World){
+    // Move the events out of the world, clearing them and avoiding a persistent borrow
+    let events = world.resource_mut::<Events<ToggleDevTool>>().drain();
+
+    // Use a resource scope to allow us to access both the dev tools registry and the rest of the world at the same time
+    world.resource_scope(|(world, dev_tools_registry: Mut<DevToolsRegistry>)|{
+        for event in events {
+            // This gives us a mutable reference to the underlying resource as a `&mut dyn ModalDevTool`
+            let Some(dev_command_metadata) = dev_tools_registry.get_command_by_name(world, event.name) else {
+                warn!("No dev tool was found for {}). Did you forget to register it?", event.name);
+                continue;
+            };
+
+            // Create a concrete instance of our dev command from the supplied string.
+            let Ok(command) = dev_command_metadata.parse_from_str(&event.0) else {
+                warn!("Could not parse the command from the supplied string");
+                continue;
+            }
+
+            // Now we can run the command directly on the `World` using `Command::apply`!
+            command.apply(world);
+        }
+    })
+}
+```
+
+Finally, we want to be able to pass in a user supplied string, parse it into a dev command and then evaluate it on the world.
+
+```rust
+#[derive(Event)]
+struct DevCommandInput(String);
```

Comment:
This doesn't seem to be used in this example:
```suggestion
```
I do wonder what the motivation behind adding 
```rust
    fn parse_from_str(string: &str) -> Result<Self, DevToolParseError> {
        Err(DevToolParseError::NotImplemented)
    }
```
to the modal dev tool interface is